resourcemanager/taskgroup/task_group.go

```diff
@@ -0,0 +1,223 @@
+}
+
+// ContextWithSchedInfo is to create a new context with task group
+func ContextWithSchedInfo(ctx context.Context) context.Context {
+	tc := fromContext(ctx)
+	if tc == nil {
+		panic("no TaskContext found in current context!")
+	}
+	return NewContext(ctx, tc.tg)
+}
+
+// CheckPoint is to check point
+func CheckPoint(ctx context.Context) {
+	tc := fromContext(ctx)
+	if tc == nil {
+		return
+	}
+
+	tg := tc.tg
+	tg.mu.RLock()
+	scheduling := tg.mu.tasks != nil
+	tg.mu.RUnlock()
+
+	if !scheduling {
+		runningTime := grunning.Time()
+		elapse := runningTime - tc.lastCheck
+		tc.lastCheck = runningTime
+		cpuTime := tg.cpuTime.Add(elapse)
+		if cpuTime < 20*time.Millisecond {
+			return
+		}
+	}
+
+	tc.wg.Add(1)
+	s.ch <- tc
+	tc.wg.Wait()
+}
+
+type sched struct {
+	ch     chan *TaskContext
+	pq     *queue.PriorityQueue[*TaskGroup]
+	isStop atomic.Bool
+}
+
+func (s *sched) stop() {
+	s.isStop.Store(true)
+}
+
+var s = sched{
+	ch: make(chan *TaskContext, 4096),
+	pq: queue.NewPriorityQueue[*TaskGroup](200, func(a *TaskGroup, b *TaskGroup) bool {
+		return a.startTime.Before(b.startTime)
+	}),
+}
+
+// StopScheduler is to stop the scheduler
+func StopScheduler() {
+	s.stop()
+}
+
+// Scheduler is to schedule the tasks
+func Scheduler() {
+	tg := NewTaskGroup()
+	ctx = NewContext(context.Background(), tg)
+	lastTime := time.Now()
+	const rate = 10
+	capacity := 200 * time.Millisecond
+	tokens := capacity
+	for {
+		cp := <-s.ch
+		tg := cp.tg
+		tg.mu.Lock()
+		if tg.mu.tasks == nil {
+			s.pq.Enqueue(tg)
+		}
+		cp.next = tg.mu.tasks
+		tg.mu.tasks = cp
+		tg.mu.Unlock()
+
+		// A token bucket algorithm to limit the total cpu usage.
+		//
+		// When a task group enqueue, it means a 20ms cpu time elapsed,
+		// assume that the elapsed wall time is 100ms, it means the CPU usage is 20ms/100ms = 20%
+		// assume that the elapsed wall time is 20ms, it means the CPU usage is 20ms/20ms = 100%
+		// assume that the elapsed wall time is 5ms, it means the CPU usage is 20ms/5ms = 400%
+		// The last case can happen in a multi-core environment.
+		//
+		// == How to decide the token generate rate? ==
+		// Say, we want to control the CPU usage at 80%, there are 10 cores.
+		// GC takes 25% of CPU resource, reserve that for it, we have actually 750% available in total.
+		// 750% * 80% = 600%
+		// CPU usage = cpu time / wall time, so the rate is: every 100ms, we can have 600ms cpu time,
+		// 600ms / 20ms = 30, or we can say: every 100ms, we can have 30 token generated.
+
+		now := time.Now()
+		elapse := now.Sub(lastTime)
+		lastTime = now
+
+		// refill tokens
+		tokens += rate * elapse
+		if tokens > capacity {
+			tokens = capacity
+		}
+
+		// == How to decide the priority of a task group? ==
+		// If a task group A is swap-in to the queue, and its last running time is [startA, endA], the corresponding cpu time is 20ms.
+		// It means the CPU resource usage rate for A is: 20ms / [now - startA], because [endA, now] does not take any CPU resource.
```

Comment:
Is `20ms / [now - startA]` not supposed to be `20ms / [endA - startA]`?

Reply:
It is not a typo. `20ms / [now - startA]`, rather than `20ms / [endA - startA]`

CPU usage = on-cpu time / duration (from wall time's perspective)
and `now` is the wall time and duration is `[now - startA]`.
**`endA` is the time when the task been swap-in to the scheduler**